src/main/java/net/dries007/tfc/util/interaction/InteractionManager.java

```diff
@@ -177,6 +181,19 @@
             return EnumActionResult.FAIL;
         });
 
+        RIGHT_CLICK_ACTIONS.put(stack -> Items.GLASS_BOTTLE.equals(stack.getItem()), (worldIn, playerIn, handIn) -> {
+            RayTraceResult traceResult = Helpers.rayTrace(worldIn, playerIn, true);
+            if (traceResult == null)
+            {
+                return EnumActionResult.PASS;
+            }
+
+            Block rayTracedBlock = worldIn.getBlockState(traceResult.getBlockPos()).getBlock();
+            boolean isTfcFluid = FluidsTFC.getAllWrappers().stream()
+                .anyMatch(fluidWrapper -> fluidWrapper.get().getBlock().equals(rayTracedBlock));
+
+            return isTfcFluid ? EnumActionResult.FAIL : EnumActionResult.PASS;
```

Comment:
The condition to ignore the action here should be if the fluid matches the check in `ItemGlassBottle` (which is a `material == Material.WATER` check, and the fluid is not equal to water. In addition, if that is the case, we want to try and fill the bottle using a fluid handler (i.e. call to `FluidUtil#tryFillContainerAnyStow`), in the event that something adds a proper fluid handler to the bottle. Otherwise, the action will fail.

If the fluid isn't `Material.WATER`, we need to manually call to `Item#onRightClick`, as it wouldn't get called otherwise.

Reply:
I don't think that spreading calls to original logic across multiple components will serve well.
We already have code to call original logic in `ClientInteractionManager` and `ServerInteractionManager` for `RightClickBlock` event. 
I suggest to move `Item.onItemRightClick` to interaction managers. As of now there is no special handling for `Server/ClientInteractionManager.processRightClickItem` it simply delegates execution to IRightClickItemAction it will not call original logic on `EnumActionResult.PASS`.

Reply:
The reason we don't just hook `RightClickBlock` event is because then we end up intercepting the right click event before `onBlockActivated` is called, leading to inconsistent placement effects. (For instance, see #414) This intercept at lowest priority, re-call all logic, then cancel the event regardless is done to *only* intercept `Item#onRightClick`

Not calling back to vanilla logic on pass is intentional - it has already been called (or supersceeded by an action) In the case with glass bottles, manually calling `ItemGlassBottle#onRightClick` is simply an event based workaround to overriding the method and calling super.